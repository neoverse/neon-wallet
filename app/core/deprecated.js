// @flow
import { get } from 'lodash'

import blockHeightActions from '../actions/blockHeightActions'
import { ID as NETWORK_ID } from '../actions/networkActions'
import { ID as SETTINGS_ID } from '../actions/settingsActions'
import { getNetworks } from '../core/networks'

const PREFIX = 'api'

// TODO: Module functions that inspect state such as `doClaimNotify` should be refactored into
//       higher-order components that expose function props for performing these actions.  Data that
//       is normally provided by functions like `getNetwork` can instead be retrieved via the
//       `withData` higher-order component.

export const getNetworkId = (state: Object) => {
  return get(state, `${PREFIX}.${NETWORK_ID}.data`)
}

export const getNetwork = (state: Object) => {
  const selectedNetworkId = getNetworkId(state)
  const networks = getNetworks()
  const networkItem = networks.find(({ id, value }) => id === selectedNetworkId) || networks[0]

  return networkItem.network
}

export const getNetworkById = (networkId: string) => {
  const networks = getNetworks()
  const networkItem = networks.find(({ id, value }) => id === networkId) || networks[0]
  return networkItem.network
}

export const getCurrency = (state: Object) => {
  return get(state, `${PREFIX}.${SETTINGS_ID}.data.currency`)
}

export const getTokensForNetwork = (state: Object) => {
  const selectedNetworkId = getNetworkId(state)
  const allTokens = get(state, `${PREFIX}.${SETTINGS_ID}.data.tokens`)
  return allTokens.filter(({ networkId }) => networkId === selectedNetworkId)
}

export const syncBlockHeight = (state: Object) => {
  return blockHeightActions.request({ networkId: getNetworkId(state) })
}
