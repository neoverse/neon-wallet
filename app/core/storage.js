import storage from 'electron-json-storage'

export const getStorage = async (key) => {
  return storage.get(key)
}

export const setStorage = async (key, value) => {
  return storage.set(key, value)
}
