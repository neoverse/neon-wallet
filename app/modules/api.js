// @flow
import { get } from 'lodash'

import {
  ACTION_REQUEST,
  ACTION_SUCCESS,
  ACTION_FAILURE,
  ACTION_RESET,
  ACTION_CANCEL,
  type Data,
  type Error,
  type ActionState
} from '../values/api'

import {
  INITIAL,
  LOADING,
  LOADED,
  FAILED,
  type ProgressState
} from '../values/state'

type State = {
  state: ProgressState,
  rollbackState: ProgressState | null,
  data: Data,
  error: Error
}

const initialState: State = {
  state: INITIAL,
  rollbackState: null,
  data: null,
  error: null
}

function reduceRequest (state: State = initialState, actionState: ActionState): Object {
  const requestType = get(actionState, 'meta.type')

  switch (requestType) {
    case ACTION_REQUEST:
      return { ...state, state: LOADING, rollbackState: state.state }
    case ACTION_SUCCESS:
      return { ...state, state: LOADED, rollbackState: LOADED, data: actionState.payload }
    case ACTION_FAILURE:
      return { ...state, state: FAILED, rollbackState: FAILED, error: actionState.payload }
    case ACTION_RESET:
      return { ...state, ...initialState }
    case ACTION_CANCEL:
      return { ...state, state: state.rollbackState }
    default:
      return state
  }
}

export default function apiReducer (state: Object = {}, actionState: ActionState): Object {
  const requestId = get(actionState, 'meta.id')
  const requestType = get(actionState, 'meta.type')

  if (requestType) {
    return {
      ...state,
      [requestId]: reduceRequest(state[requestId], actionState)
    }
  } else {
    return state
  }
}
