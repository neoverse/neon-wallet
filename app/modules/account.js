// @flow
import { wallet } from 'neon-js'

import { showErrorNotification, showInfoNotification, hideNotification } from './notifications'

import commNode from '../ledger/ledger-comm-node'
import { ledgerNanoSCreateSignatureAsync } from '../ledger/ledgerNanoS'

import { upgradeNEP6AddAddresses } from '../core/account'
import { validatePassphraseLength } from '../core/wallet'
import { BIP44_PATH, ROUTES, FINDING_LEDGER_NOTICE } from '../core/constants'
import asyncWrap from '../core/asyncHelper'

// Constants
export const LOGOUT = 'LOGOUT'
export const HARDWARE_DEVICE_INFO = 'HARDWARE_DEVICE_INFO'
export const HARDWARE_PUBLIC_KEY_INFO = 'HARDWARE_PUBLIC_KEY_INFO'
export const HARDWARE_PUBLIC_KEY = 'HARDWARE_PUBLIC_KEY'
export const HARDWARE_LOGIN = 'HARDWARE_LOGIN'

// Actions
export function logout () {
  return {
    type: LOGOUT
  }
}

export function hardwareDeviceInfo (hardwareDeviceInfo: string) {
  return {
    type: HARDWARE_DEVICE_INFO,
    payload: { hardwareDeviceInfo }
  }
}

export function hardwarePublicKeyInfo (hardwarePublicKeyInfo: ?string) {
  return {
    type: HARDWARE_PUBLIC_KEY_INFO,
    payload: { hardwarePublicKeyInfo }
  }
}

export function hardwarePublicKey (publicKey: ?string) {
  return {
    type: HARDWARE_PUBLIC_KEY,
    payload: { publicKey }
  }
}

export function isHardwareLogin (isHardwareLogin: boolean) {
  return {
    type: HARDWARE_LOGIN,
    payload: { isHardwareLogin }
  }
}

// Reducer that manages account state (account now = private key)
export const ledgerNanoSGetInfoAsync = () => async (dispatch: DispatchType) => {
  const dispatchError = (message: string, deviceInfoMsg: boolean = true) => {
    dispatch(isHardwareLogin(false))
    dispatch(hardwarePublicKey(null))
    if (deviceInfoMsg) {
      dispatch(hardwarePublicKeyInfo(null))
      return dispatch(hardwareDeviceInfo(message))
    } else {
      return dispatch(hardwarePublicKeyInfo(message))
    }
  }
  dispatch(hardwareDeviceInfo(FINDING_LEDGER_NOTICE))
  let [err, result] = await asyncWrap(commNode.list_async())
  if (err) {
    return dispatchError(`Finding USB Error: ${err}. ${FINDING_LEDGER_NOTICE}`)
  }
  if (result.length === 0) {
    return dispatchError(`USB Failure: No device found. ${FINDING_LEDGER_NOTICE}`)
  } else {
    let [err, comm] = await asyncWrap(commNode.create_async())
    if (err) {
      return dispatchError(`Finding USB Error: ${err}. ${FINDING_LEDGER_NOTICE}`)
    }

    const deviceInfo = comm.device.getDeviceInfo()
    comm.device.close()
    dispatch(hardwareDeviceInfo(`Found USB ${deviceInfo.manufacturer} ${deviceInfo.product}`))
  }
  [err, result] = await asyncWrap(commNode.list_async())
  if (result.length === 0) {
    return dispatchError('Hardware Device Error. Login to NEO App and try again', false)
  } else {
    let [err, comm] = await asyncWrap(commNode.create_async())
    if (err) {
      console.log(`Public Key Comm Init Error: ${err}`)
      return dispatchError('Hardware Device Error. Login to NEO App and try again', false)
    }

    let message = Buffer.from(`8004000000${BIP44_PATH}`, 'hex')
    const validStatus = [0x9000]
    let [error, response] = await asyncWrap(comm.exchange(message.toString('hex'), validStatus))
    if (error) {
      comm.device.close() // NOTE: do we need this close here - what about the other errors that do not have it at the moment
      if (error === 'Invalid status 28160') {
        return dispatchError('NEO App does not appear to be open, request for private key returned error 28160.', false)
      } else {
        console.log(`Public Key Comm Messaging Error: ${error}`)
        return dispatchError('Hardware Device Error. Login to NEO App and try again', false)
      }
    }
    comm.device.close()
    dispatch(isHardwareLogin(true))
    dispatch(hardwarePublicKey(response.substring(0, 130)))
    return dispatch(hardwarePublicKeyInfo('Success. NEO App Found on Hardware Device. Click Button Above to Login'))
  }
}

// State Getters
export const getWIF = (state: Object) => state.account.wif
export const getAddress = (state: Object) => state.account.address
export const getLoggedIn = (state: Object) => state.account.loggedIn
export const getRedirectUrl = (state: Object) => state.account.redirectUrl
export const getAccounts = (state: Object) => state.account.accounts
export const getSigningFunction = (state: Object) => state.account.signingFunction
export const getPublicKey = (state: Object) => state.account.publicKey
export const getHardwareDeviceInfo = (state: Object) => state.account.hardwareDeviceInfo
export const getHardwarePublicKeyInfo = (state: Object) => state.account.hardwarePublicKeyInfo
export const getIsHardwareLogin = (state: Object) => state.account.isHardwareLogin

const initialState = {
  wif: null,
  address: null,
  loggedIn: false,
  redirectUrl: null,
  accounts: [],
  signingFunction: null,
  publicKey: null,
  isHardwareLogin: false,
  hardwareDeviceInfo: null,
  hardwarePublicKeyInfo: null
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case LOGOUT:
      return {
        ...state,
        wif: null,
        address: null,
        loggedIn: false,
        signingFunction: null,
        publicKey: null,
        isHardwareLogin: false
      }
    case HARDWARE_DEVICE_INFO:
      const { hardwareDeviceInfo } = action.payload
      return {
        ...state,
        hardwareDeviceInfo
      }
    case HARDWARE_LOGIN:
      const { isHardwareLogin } = action.payload
      return {
        ...state,
        isHardwareLogin
      }
    case HARDWARE_PUBLIC_KEY_INFO:
      const { hardwarePublicKeyInfo } = action.payload
      return {
        ...state,
        hardwarePublicKeyInfo
      }
    case HARDWARE_PUBLIC_KEY:
      const { publicKey } = action.payload
      return {
        ...state,
        publicKey
      }
    default:
      return state
  }
}
