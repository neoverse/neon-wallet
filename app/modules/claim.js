// @flow
import { api, type Claims } from 'neon-js'
import { map, reduce } from 'lodash'

import {
  showErrorNotification,
  showSuccessNotification,
  showInfoNotification
} from './notifications'
import {
  getNetwork,
  getWIF,
  getAddress,
  getSigningFunction,
  getPublicKey,
  getIsHardwareLogin,
  getNEO
} from '../core/deprecated'
import { toBigNumber, toNumber } from '../core/math'
import { ASSETS } from '../core/constants'
import { FIVE_MINUTES_MS } from '../core/time'
import poll from '../util/poll'

const POLL_ATTEMPTS = 50
const POLL_FREQUENCY = 5000

const fetchClaims = async ({ net, address }) => {
  api.setApiSwitch(1)
  const response = await api.loadBalance(api.getClaimsFrom, { net, address })
  api.setApiSwitch(0)
  const { claims } = response.claims
  return map(claims, 'claim')
}

const calculateClaimableAmount = (claims: Claims) => {
  return reduce(claims, (sum, claim) => claim.plus(sum), 0).toString()
}

const getClaimableAmount = async ({ net, address }) => {
  const claims = await fetchClaims({ net, address })
  return calculateClaimableAmount(claims)
}

const updateClaimableAmount = async ({ net, address, publicKey, privateKey, signingFunction, balance }) => {
  const { response } = await api.sendAsset({
    net,
    address,
    publicKey,
    privateKey,
    signingFunction,
    intents: api.makeIntent({ [ASSETS.NEO]: toNumber(balance) }, address)
  })

  if (!response.result || !response.txid) {
    throw new Error('Rejected by RPC server.')
  }

  return response.result.response
}

const pollForUpdatedClaimableAmount = async ({ net, address, claimableAmount }) => {
  return poll(async () => {
    const updatedClaimableAmount = await getClaimableAmount({ net, address })

    if (toBigNumber(updatedClaimableAmount).eq(claimableAmount)) {
      throw new Error('Waiting for updated claims.')
    }

    return updatedClaimableAmount
  }, { attempts: POLL_ATTEMPTS, frequency: POLL_FREQUENCY })
}

const getUpdatedClaimableAmount = async ({ net, address, balance, publicKey, privateKey, signingFunction }) => {
  const claimableAmount = await getClaimableAmount({ net, address })

  if (toBigNumber(balance).eq(0)) {
    return claimableAmount
  } else {
    await updateClaimableAmount({ net, address, balance, publicKey, privateKey, signingFunction })
    return pollForUpdatedClaimableAmount({ net, address, claimableAmount })
  }
}

export const doGasClaim = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const address = getAddress(state)
  const net = getNetwork(state)
  const balance = getNEO(state)
  const publicKey = getPublicKey(state)
  const privateKey = getWIF(state)
  const signingFunction = getSigningFunction(state)
  const isHardwareClaim = getIsHardwareLogin(state)

  dispatch(disableClaim(true))

  if (isHardwareClaim) {
    dispatch(showInfoNotification({ message: 'Please sign transaction 1 of 2 on hardware device.' }))
  } else {
    dispatch(showInfoNotification({ message: 'Calculating claimable GAS...' }))
  }

  // step 1: update available claims
  try {
    await getUpdatedClaimableAmount({ net, address, balance, publicKey, privateKey, signingFunction })
  } catch (err) {
    dispatch(disableClaim(false))
    dispatch(showErrorNotification({ message: `Calculating claimable GAS failed: ${err.message}` }))
    return
  }

  if (isHardwareClaim) {
    dispatch(showInfoNotification({ message: 'Please sign transaction 2 of 2 on hardware device.' }))
  } else {
    dispatch(showInfoNotification({ message: 'Claiming GAS...' }))
  }

  // step 2: send claim request
  try {
    api.setApiSwitch(1)
    const { response } = await api.claimGas({ net, address, publicKey, privateKey, signingFunction })
    api.setApiSwitch(0)

    if (!response.result) {
      throw new Error('Rejected by RPC server.')
    }
  } catch (err) {
    dispatch(disableClaim(false))
    dispatch(showErrorNotification({ message: `Claiming GAS failed: ${err.message}` }))
    return
  }

  dispatch(showSuccessNotification({
    message: 'Claim was successful! Your balance will update once the blockchain has processed it.'
  }))
  setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
}

// Constants
export const DISABLE_CLAIM = 'DISABLE_CLAIM'

// Actions
export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

// State Getters
export const getDisableClaimButton = (state: Object) => state.claim.disableClaimButton

const initialState = {
  disableClaimButton: false
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    default:
      return state
  }
}
