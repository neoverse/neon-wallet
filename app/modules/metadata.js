// @flow
import { createSelector } from 'reselect'
import { isNil } from 'lodash'
import axios from 'axios'
import { api } from 'neon-js'
import storage from 'electron-json-storage'

import { showWarningNotification, showErrorNotification } from './notifications'
import { setCurrency } from './price'

import { getNetwork, getNetworkId } from '../core/deprecated'
import {
  EXPLORERS,
  NEON_WALLET_RELEASE_LINK,
  NOTIFICATION_POSITIONS,
  TOKENS,
  TOKENS_TEST,
  MAIN_NETWORK_ID,
  TEST_NETWORK_ID
} from '../core/constants'
import asyncWrap from '../core/asyncHelper'

import { version } from '../../package.json'

// Constants
export const SET_EXPLORER = 'SET_EXPLORER'
export const SET_USER_GENERATED_TOKENS = 'SET_USER_GENERATED_TOKENS'

// Actions
export const setBlockExplorer = (blockExplorer: ExplorerType) => async (
  dispatch: DispatchType
) => {
  storage.get('settings', (errorReading, settingsObj) => {
    if (errorReading) {
      dispatch(
        showErrorNotification({ message: 'error grabbing data from storage' })
      )
    }
    storage.set(
      'settings',
      {
        ...settingsObj,
        blockExplorer
      },
      saveError => {
        if (saveError) {
          dispatch(
            showErrorNotification({
              message: 'error saving new block explorer in storage'
            })
          )
        }
      }
    )
  })
  return dispatch({
    type: SET_EXPLORER,
    payload: { blockExplorer }
  })
}

export const setUserGeneratedTokens = (tokens: Array<TokenItemType>) => async (
  dispatch: DispatchType
) => {
  storage.get('settings', (errorReading, settingsObj) => {
    if (errorReading) {
      dispatch(
        showErrorNotification({ message: 'error grabbing data from storage' })
      )
    }
    const userGeneratedTokens = tokens.filter(
      (token: TokenItemType) => token.isUserGenerated
    )
    storage.set(
      'settings',
      {
        ...settingsObj,
        tokens: userGeneratedTokens
      },
      saveError => {
        if (saveError) {
          dispatch(
            showErrorNotification({
              message: 'error saving new tokens in storage'
            })
          )
        }
      }
    )
  })
  return dispatch({
    type: SET_USER_GENERATED_TOKENS,
    payload: { tokens }
  })
}

export const checkVersion = () => async (
  dispatch: DispatchType,
  getState: GetStateType
) => {
  const state = getState()
  const net = getNetwork(state)
  const apiEndpoint = api.neonDB.getAPIEndpoint(net)

  const [err, res] = await asyncWrap(axios.get(`${apiEndpoint}/v2/version`))
  const shouldUpdate = res && res.data && res.data.version !== version
  if (err || shouldUpdate) {
    const link = `<a href='${NEON_WALLET_RELEASE_LINK}' target='_blank' class="notification-link">${NEON_WALLET_RELEASE_LINK}</a>`
    const message = err
      ? `Error checking wallet version! Please make sure you have downloaded the latest version: ${link}`
      : `Your wallet is out of date! Please download the latest version from ${link}`
    return dispatch(
      showWarningNotification({
        message,
        autoDismiss: 0,
        stack: true,
        position: NOTIFICATION_POSITIONS.BOTTOM_CENTER
      })
    )
  }
}

export const initSettings = () => async (dispatch: DispatchType) => {
  // eslint-disable-next-line
  storage.get('settings', (error, settings) => {
    if (!isNil(settings.blockExplorer)) {
      dispatch(setBlockExplorer(settings.blockExplorer))
    }

    if (!isNil(settings.currency)) {
      dispatch(setCurrency(settings.currency))
    }

    if (!isNil(settings.tokens)) {
      dispatch(setUserGeneratedTokens(settings.tokens))
    }
  })
}

// state getters
export const getBlockExplorer = (state: Object) => state.metadata.blockExplorer
export const getAllTokens = (state: Object) => state.metadata.tokens

// computed state getters

export const getTokensForNetwork = createSelector(
  getAllTokens,
  getNetworkId,
  (allTokens, selectedNetworkId) =>
    allTokens.filter(({ networkId }) => networkId === selectedNetworkId)
)

const generatePredfinedTokens = (): Array<TokenItemType> => {
  const tokens = []
  let id = 1

  const getTokenEntry = (
    scriptHash: string,
    networkId: string
  ) => ({
    id: `${id++}`,
    scriptHash,
    networkId,
    isUserGenerated: false
  })

  Object.keys(TOKENS).forEach(symbol => {
    const scriptHash = TOKENS[symbol]
    tokens.push(getTokenEntry(scriptHash, MAIN_NETWORK_ID))
  })

  Object.keys(TOKENS_TEST).forEach(symbol => {
    const scriptHash = TOKENS_TEST[symbol]
    tokens.push(getTokenEntry(scriptHash, TEST_NETWORK_ID))
  })

  return tokens
}

const initialState = {
  tokens: generatePredfinedTokens(),
  blockExplorer: EXPLORERS.NEO_TRACKER
}

export default (state: Object = initialState, action: ReduxAction) => {
  switch (action.type) {
    case SET_EXPLORER:
      const { blockExplorer } = action.payload
      return {
        ...state,
        blockExplorer
      }
    case SET_USER_GENERATED_TOKENS:
      const { tokens } = action.payload
      return {
        ...state,
        tokens: [
          // keep existing non user generated tokens
          ...state.tokens.filter(token => !token.isUserGenerated),
          ...tokens.filter(token => token.isUserGenerated)
        ]
      }
    default:
      return state
  }
}
