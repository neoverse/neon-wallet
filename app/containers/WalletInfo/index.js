// @flow
import { connect, type MapStateToProps } from 'react-redux'
import { bindActionCreators } from 'redux'
import { compose } from 'recompose'

import withActions from '../../hocs/api/withActions'
import withNetworkData from '../../hocs/withNetworkData'
import withCurrencyData from '../../hocs/withCurrencyData'
import { updateSettingsActions } from '../../actions/settingsActions'
import { getNetworks } from '../../core/networks'
import { showErrorNotification, showSuccessNotification } from '../../modules/notifications'
import { getAddress } from '../../modules/account'
import { loadWalletData, getNEO, getGAS, getTokenBalances } from '../../modules/wallet'
import { getNEOPrice, getGASPrice } from '../../modules/price'
import { showModal } from '../../modules/modal'
import { participateInSale, oldParticipateInSale } from '../../modules/sale'

import WalletInfo from './WalletInfo'

const mapStateToProps: MapStateToProps<*, *, *> = (state: Object) => ({
  NEO: getNEO(state),
  GAS: getGAS(state),
  address: getAddress(state),
  neoPrice: getNEOPrice(state),
  gasPrice: getGASPrice(state),
  tokenBalances: getTokenBalances(state),
  networks: getNetworks()
})

const actionCreators = {
  loadWalletData,
  showErrorNotification,
  showSuccessNotification,
  showModal,
  participateInSale,
  oldParticipateInSale
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

const mapActionsToProps = (actions) => ({
  setUserGeneratedTokens: (tokens) => actions.request({ tokens })
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withNetworkData(),
  withCurrencyData('currencyCode'),
  withActions(updateSettingsActions, mapActionsToProps)
)(WalletInfo)
