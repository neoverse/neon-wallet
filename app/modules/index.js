// @flow
import { combineReducers } from 'redux'
import api from './api'
import account from './account'
import generateWallet from './generateWallet'
import transactions from './transactions'
import wallet from './wallet'
import claim from './claim'
import dashboard from './dashboard'
import notifications from './notifications'
import modal from './modal'
import price from './price'
import addressBook from './addressBook'
import sale from './sale'

export default combineReducers({
  api,
  account,
  generateWallet,
  wallet,
  transactions,
  dashboard,
  claim,
  notifications,
  modal,
  price,
  addressBook,
  sale
})
