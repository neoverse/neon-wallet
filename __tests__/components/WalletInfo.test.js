import React from 'react'
import * as neonjs from 'neon-js'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { mount, shallow } from 'enzyme'

import {
  SET_TRANSACTION_HISTORY,
  SET_BALANCE
} from '../../app/modules/wallet'
import { SHOW_NOTIFICATION } from '../../app/modules/notifications'
import { LOADING_TRANSACTIONS } from '../../app/modules/transactions'

import { DEFAULT_CURRENCY_CODE, MAIN_NETWORK_ID } from '../../app/core/constants'
import { LOADED } from '../../app/values/state'

import WalletInfo from '../../app/containers/WalletInfo'

// TODO research how to move the axios mock code which is repeated in NetworkSwitch to a helper or config file
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import { version } from '../../package.json'

const axiosMock = new MockAdapter(axios)
axiosMock
  .onGet('http://testnet-api.wallet.cityofzion.io/v2/version')
  .reply(200, { version })
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/NEO/?convert=USD')
  .reply(200, [{ price_usd: 24.5 }])
axiosMock
  .onGet('https://api.coinmarketcap.com/v1/ticker/GAS/?convert=USD')
  .reply(200, [{ price_usd: 18.2 }])

jest.mock('electron', () => ({
  app: {
    getPath: () => {
      return 'C:\\tmp\\mock_path'
    }
  }
}))
jest.useFakeTimers()

jest.unmock('qrcode')
import QRCode from 'qrcode/lib/browser' // eslint-disable-line
QRCode.toCanvas = jest.fn()

const initialState = {
  api: {
    NETWORK: {
      batch: false,
      state: LOADED,
      data: MAIN_NETWORK_ID
    }
  },
  account: {
    address: 'ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw'
  },
  metadata: {
  },
  wallet: {
    NEO: '100001',
    GAS: '1000.0001601',
    tokenBalances: []
  },
  price: {
    NEO: 25.48,
    GAS: 18.1,
    currency: DEFAULT_CURRENCY_CODE
  },
  claim: {
    claimAmount: 0.5
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<WalletInfo store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <WalletInfo />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('WalletInfo', () => {
  test('renders without crashing', done => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
    done()
  })
  test('correctly renders data from state', done => {
    const { wrapper } = setup(initialState, false)

    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')

    const expectedNeoWalletValue = '2,548,025.48'
    const expectedGasWalletValue = '18,100.00'
    const expectedWalletValue = '2,566,125.48'
    const neoField = wrapper.find('.amountNeo')
    const gasField = wrapper.find('.amountGas')

    expect(neoWalletValue.text()).toEqual(`$${expectedNeoWalletValue} USD`)
    expect(gasWalletValue.text()).toEqual(`$${expectedGasWalletValue} USD`)
    expect(walletValue.text()).toEqual(`Total $${expectedWalletValue} USD`)
    expect(neoField.text()).toEqual('100,001')
    // TODO: Test the GAS tooltip value, this is testing the display value, truncated to 4 decimals
    expect(gasField.text()).toEqual('1,000.0002')
    done()
  })
  test('refreshBalance is getting called on click', async () => {
    const { wrapper, store } = setup(initialState, false)

    wrapper.find('.refreshBalance').simulate('click')

    await Promise.resolve('Pause')
      .then()
      .then()
      .then()
    jest.runAllTimers()
    const actions = store.getActions()
    expect(actions.length).toEqual(7)

    expect(actions[0]).toEqual({
      type: LOADING_TRANSACTIONS,
      payload: {
        isLoadingTransactions: true
      }
    })
    expect(actions[1]).toEqual({
      type: LOADING_TRANSACTIONS,
      payload: {
        isLoadingTransactions: false
      }
    })
    expect(actions[2]).toEqual({
      type: SET_TRANSACTION_HISTORY,
      payload: {
        transactions: []
      }
    })
    expect(actions[4]).toEqual({
      type: SET_BALANCE,
      payload: {
        NEO: '1',
        GAS: '1'
      }
    })
    // TODO fix this to capture the notifications as well
    // expect(actions[6]).toEqual({
    //   type: HIDE_NOTIFICATIONS,
    //   payload: {
    //     dismissible: true,
    //     position: DEFAULT_POSITION
    //   }
    // })
    // expect(actions[7]).toEqual({
    //   type: SHOW_NOTIFICATION,
    //   payload: expect.objectContaining({
    //     message: 'Received latest blockchain information.',
    //     level: NOTIFICATION_LEVELS.SUCCESS
    //   })
    // })
  })
  test('correctly renders data from state with non-default currency', done => {
    const testState = {
      ...initialState,
      price: { NEO: 1.11, GAS: 0.55, currency: 'eur' }
    }
    const { wrapper } = setup(testState, false)

    const neoWalletValue = wrapper.find('.neoWalletValue')
    const gasWalletValue = wrapper.find('.gasWalletValue')
    const walletValue = wrapper.find('.walletTotal')

    const expectedNeoWalletValue = '111,001.11'
    const expectedGasWalletValue = '550.00'
    const expectedWalletValue = '111,551.11'

    expect(neoWalletValue.text()).toEqual(`€${expectedNeoWalletValue} EUR`)
    expect(gasWalletValue.text()).toEqual(`€${expectedGasWalletValue} EUR`)
    expect(walletValue.text()).toEqual(`Total €${expectedWalletValue} EUR`)

    done()
  })
  test('network error is shown with connectivity error', async () => {
    neonjs.api.neonDB.getBalance = jest.fn(() => {
      return new Promise((resolve, reject) => {
        reject(new Error())
      })
    })
    const { wrapper, store } = setup(initialState, false)
    wrapper.find('.refreshBalance').simulate('click')

    jest.runAllTimers()
    await Promise.resolve('Pause')
      .then()
      .then()
      .then()
      .then()

    const actions = store.getActions()
    const notifications = []
    actions.forEach(action => {
      if (action.type === SHOW_NOTIFICATION) {
        notifications.push(action)
      }
    })

    // let's make sure the last notification show was an error.
    expect(notifications.pop().payload.level).toEqual('error')
  })
})
