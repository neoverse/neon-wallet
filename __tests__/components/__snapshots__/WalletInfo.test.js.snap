// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletInfo renders without crashing 1`] = `
<Connect(withData(mapProps(Connect(withData(Connect(withActions(WalletInfo)))))))
  GAS="1000.0001601"
  NEO="100001"
  address="ANqUrhv99rwCiFTL6N1An9NH5UVkPYxTuw"
  gasPrice={18.1}
  loadWalletData={[Function]}
  neoPrice={25.48}
  networks={
    Array [
      Object {
        "id": "1",
        "label": "MainNet",
        "network": "MainNet",
      },
      Object {
        "id": "2",
        "label": "TestNet",
        "network": "TestNet",
      },
    ]
  }
  oldParticipateInSale={[Function]}
  participateInSale={[Function]}
  showErrorNotification={[Function]}
  showModal={[Function]}
  showSuccessNotification={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  tokenBalances={Array []}
/>
`;
